src/resolution/graph.rs

```diff
@@ -968,208 +1040,191 @@ impl<'a, TNpmRegistryApi: NpmRegistryApi>
 
   pub async fn resolve_pending(&mut self) -> Result<(), NpmResolutionError> {
     // go down through the dependencies by tree depth
-    while let Some(parent_path) = self.pending_unresolved_nodes.pop_front() {
-      let (parent_nv, child_deps) = {
-        let node_id = parent_path.node_id();
-        if self.graph.nodes.get(&node_id).unwrap().no_peers {
-          // We can skip as there's no reason to analyze this graph segment further.
-          continue;
+    while !self.pending_unresolved_nodes.is_empty() {
+      while let Some(parent_path) = self.pending_unresolved_nodes.pop_front() {
+        self.resolve_next_pending(parent_path).await?;
+      }
+
+      if !self.graph.unresolved_optional_peers.take_seen().is_empty() {
+        for (nv, node_id) in &self.graph.root_packages {
+          self
+            .pending_unresolved_nodes
+            .push_back(GraphPath::for_root(*node_id, nv.clone()));
         }
+      }
+    }
+    Ok(())
+  }
 
-        let pkg_nv = self
-          .graph
-          .resolved_node_ids
-          .get(node_id)
-          .unwrap()
-          .nv
-          .clone();
-        let deps = if let Some(deps) = self.dep_entry_cache.get(&pkg_nv) {
-          deps.clone()
-        } else {
-          // the api is expected to have cached this at this point, so no
-          // need to parallelize
-          let package_info = self.api.package_info(&pkg_nv.name).await?;
-          let version_info = package_info
-            .version_info(&pkg_nv, self.version_resolver.patch_packages)
-            .map_err(NpmPackageVersionResolutionError::VersionNotFound)?;
-          self.dep_entry_cache.store(pkg_nv.clone(), version_info)?
-        };
-
-        (pkg_nv, deps)
+  async fn resolve_next_pending(
```

Comment:
The diff here is pretty bad. It's mostly extracting this out to a new method. I'll comment on what changed.